File: spec/test_error_positions.c
#include "testparser.h"
#include <assert.h>
#include <string.h>

int main()
{
    char const * input = "a 42";
    p_context_t * context;
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert(p_parse(context) == P_SUCCESS);
    p_context_delete(context);

    input = "a\n123\na  a";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert(p_parse(context) == P_UNEXPECTED_TOKEN);
    assert(p_position(context).row == 3);
    assert(p_position(context).col == 4);
    assert(p_token(context) == TOKEN_a);
    p_context_delete(context);

    input = "12";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert(p_parse(context) == P_UNEXPECTED_TOKEN);
    assert(p_position(context).row == 1);
    assert(p_position(context).col == 1);
    assert(p_token(context) == TOKEN_num);
    p_context_delete(context);

    input = "a 12\n\nab";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert(p_parse(context) == P_UNEXPECTED_INPUT);
    assert(p_position(context).row == 3);
    assert(p_position(context).col == 2);
    p_context_delete(context);

    input = "a 12\n\na\n\n77\na   \xAA";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert(p_parse(context) == P_DECODE_ERROR);
    assert(p_position(context).row == 6);
    assert(p_position(context).col == 5);

    assert(strcmp(p_token_names[TOKEN_a], "a") == 0);
    assert(strcmp(p_token_names[TOKEN_num], "num") == 0);
    p_context_delete(context);

    return 0;
}

File: spec/test_error_positions.rs
use testparser::*;

fn main() {
    let mut c = p_context_new(b"a 42");
    assert_eq!(P_SUCCESS, p_parse(&mut c));
    p_context_delete(c);

    let mut c = p_context_new(b"a\n123\na  a");
    assert_eq!(P_UNEXPECTED_TOKEN, p_parse(&mut c));
    assert_eq!(3, p_position(&c).row);
    assert_eq!(4, p_position(&c).col);
    assert_eq!(TOKEN_a, p_token(&c));
    p_context_delete(c);

    let mut c = p_context_new(b"12");
    assert_eq!(P_UNEXPECTED_TOKEN, p_parse(&mut c));
    assert_eq!(1, p_position(&c).row);
    assert_eq!(1, p_position(&c).col);
    assert_eq!(TOKEN_num, p_token(&c));
    p_context_delete(c);

    let mut c = p_context_new(b"a 12\n\nab");
    assert_eq!(P_UNEXPECTED_INPUT, p_parse(&mut c));
    assert_eq!(3, p_position(&c).row);
    assert_eq!(2, p_position(&c).col);
    p_context_delete(c);

    let mut c = p_context_new(b"a 12\n\na\n\n77\na   \xAA");
    assert_eq!(P_DECODE_ERROR, p_parse(&mut c));
    assert_eq!(6, p_position(&c).row);
    assert_eq!(5, p_position(&c).col);
    assert_eq!("a", p_token_names[TOKEN_a as usize]);
    assert_eq!("num", p_token_names[TOKEN_num as usize]);
    p_context_delete(c);
}
